File: chapter_api_tests/2024-04/validation/test_get_api-v1-projects-{projectRef}-runs.py
import axios, { AxiosInstance, AxiosError } from 'axios';\nimport { describe, it, expect, beforeAll } from '@jest/globals';\n\ndescribe('GET /api/v1/projects/{projectRef}/runs - List project runs', () => {\n  let client: AxiosInstance;\n  const validProjectRef = 'my-valid-project';\n  const invalidProjectRef = 'non-existing-project';\n\n  beforeAll(() => {\n    // Create an Axios instance with the base URL and authorization header.\n    // These values are loaded from environment variables (API_BASE_URL and API_AUTH_TOKEN).\n    client = axios.create({\n      baseURL: process.env.API_BASE_URL,\n      headers: {\n        Authorization: `Bearer ${process.env.API_AUTH_TOKEN}`\n      }\n    });\n  });\n\n  it('should return 200 and a valid response when called with correct parameters', async () => {\n    const params = {\n      env: 'production',\n      status: 'success',\n      limit: 5\n    };\n\n    const response = await client.get(`/api/v1/projects/${validProjectRef}/runs`, { params });\n\n    // 1. Response status and headers\n    expect(response.status).toBe(200);\n    expect(response.headers['content-type']).toContain('application/json');\n\n    // 2. Basic validation of the response body\n    //    This should match #/components/schemas/ListRunsResult in principle.\n    expect(response.data).toHaveProperty('runs');\n    expect(Array.isArray(response.data.runs)).toBe(true);\n  });\n\n  it('should return 400 (or 422) for invalid query parameter type (e.g., limit=abc)', async () => {\n    const params = { limit: 'abc' };\n\n    try {\n      await client.get(`/api/v1/projects/${validProjectRef}/runs`, { params });\n      throw new Error('Request should have failed due to invalid parameter type');\n    } catch (error) {\n      const axiosError = error as AxiosError;\n      if (axiosError.response) {\n        // The API could return 400 or 422 for invalid request payloads.\n        expect([400, 422]).toContain(axiosError.response.status);\n      } else {\n        throw error;\n      }\n    }\n  });\n\n  it('should return 401 (or 403) if no auth token is provided', async () => {\n    const unauthClient = axios.create({\n      baseURL: process.env.API_BASE_URL\n    });\n\n    try {\n      await unauthClient.get(`/api/v1/projects/${validProjectRef}/runs`);\n      throw new Error('Request should have failed due to missing token');\n    } catch (error) {\n      const axiosError = error as AxiosError;\n      if (axiosError.response) {\n        // The API may return 401 or 403 in these cases.\n        expect([401, 403]).toContain(axiosError.response.status);\n      } else {\n        throw error;\n      }\n    }\n  });\n\n  it('should return 404 if the projectRef does not exist', async () => {\n    try {\n      await client.get(`/api/v1/projects/${invalidProjectRef}/runs`);\n      throw new Error('Request should have failed due to invalid projectRef');\n    } catch (error) {\n      const axiosError = error as AxiosError;\n      if (axiosError.response) {\n        expect(axiosError.response.status).toBe(404);\n      } else {\n        throw error;\n      }\n    }\n  });\n\n  it('should return an empty list if no runs are found', async () => {\n    // Provide parameters that are unlikely to match any run\n    const params = { status: 'does-not-exist-123' };\n    const response = await client.get(`/api/v1/projects/${validProjectRef}/runs`, { params });\n\n    expect(response.status).toBe(200);\n    expect(response.data).toHaveProperty('runs');\n    expect(response.data.runs.length).toBe(0);\n  });\n\n  it('should handle cursor pagination parameters correctly', async () => {\n    // Example of testing pagination-related parameters (cursorPagination).\n    const params = { limit: 1 };\n    const response = await client.get(`/api/v1/projects/${validProjectRef}/runs`, { params });\n\n    expect(response.status).toBe(200);\n    expect(Array.isArray(response.data.runs)).toBe(true);\n    // Here, you might also check for 'nextCursor', 'prevCursor', etc., if defined in the schema.\n  });\n\n  it('should return 400 (or 422) for malformed or out-of-range date filters', async () => {\n    // Assuming runsFilterWithEnv allows date filters, e.g. createdAtBefore, createdAtAfter.\n    const params = { createdAtBefore: 'not-a-date' };\n\n    try {\n      await client.get(`/api/v1/projects/${validProjectRef}/runs`, { params });\n      throw new Error('Request should have failed due to invalid date');\n    } catch (error) {\n      const axiosError = error as AxiosError;\n      if (axiosError.response) {\n        // The API could return 400 or 422 for invalid request payloads.\n        expect([400, 422]).toContain(axiosError.response.status);\n      } else {\n        throw error;\n      }\n    }\n  });\n});\n

File: chapter_api_tests/2024-04/validation/test_get_api-v1-projects-{projectRef}-runs.ts
import axios, { AxiosError } from 'axios';

describe('GET /api/v1/projects/{projectRef}/runs', () => {
  const baseURL = process.env.API_BASE_URL || 'http://localhost:3000';
  const token = process.env.API_AUTH_TOKEN || '';
  const validProjectRef = 'my-valid-project';
  const invalidProjectRef = '!@#'; // some invalid reference
  const path = '/api/v1/projects';

  beforeAll(() => {
    if (!baseURL) {
      console.warn('API_BASE_URL is not set. Tests may fail.');
    }
  });

  describe('Input Validation', () => {
    it('should return 200 for valid query parameters', async () => {
      const url = `${baseURL}${path}/${validProjectRef}/runs?status=completed&page=1&limit=5`;
      try {
        const response = await axios.get(url, {
          headers: {
            Authorization: `Bearer ${token}`,
          },
        });
        expect(response.status).toBe(200);
        // Additional assertions about response structure could go here
      } catch (error) {
        // If there's an error, force the test to fail
        throw new Error(`Unexpected error: ${error}`);
      }
    });

    it('should return 400 or 422 for invalid query parameter types', async () => {
      const url = `${baseURL}${path}/${validProjectRef}/runs?limit=abc`; // limit is invalid type
      try {
        await axios.get(url, {
          headers: {
            Authorization: `Bearer ${token}`,
          },
        });
        fail('Expected request to fail');
      } catch (error) {
        const axiosError = error as AxiosError;
        expect([400, 422]).toContain(axiosError.response?.status);
      }
    });
  });

  describe('Response Validation', () => {
    it('should return valid JSON and status 200 for a successful request', async () => {
      const url = `${baseURL}${path}/${validProjectRef}/runs`;
      const response = await axios.get(url, {
        headers: {
          Authorization: `Bearer ${token}`,
        },
      });
      expect(response.status).toBe(200);
      expect(response.headers['content-type']).toContain('application/json');
      // Here we can do partial schema validation:
      expect(response.data).toHaveProperty('runs');
      expect(Array.isArray(response.data.runs)).toBe(true);
    });

    it('should handle invalid projectRef gracefully', async () => {
      const url = `${baseURL}${path}/${invalidProjectRef}/runs`;
      try {
        await axios.get(url, {
          headers: {
            Authorization: `Bearer ${token}`,
          },
        });
        fail('Expected request to fail due to invalid projectRef');
      } catch (error) {
        const axiosError = error as AxiosError;
        // The API might return 400 or 404 if the projectRef is not valid
        expect([400, 404]).toContain(axiosError.response?.status);
      }
    });
  });

  describe('Response Headers Validation', () => {
    it('should have correct Content-Type header', async () => {
      const url = `${baseURL}${path}/${validProjectRef}/runs`;
      const response = await axios.get(url, {
        headers: {
          Authorization: `Bearer ${token}`,
        },
      });
      expect(response.headers['content-type']).toContain('application/json');
    });
  });

  describe('Edge Case & Limit Testing', () => {
    it('should return empty array if no runs are found', async () => {
      const emptyProjectRef = 'project-with-no-runs';
      const url = `${baseURL}${path}/${emptyProjectRef}/runs`;
      const response = await axios.get(url, {
        headers: {
          Authorization: `Bearer ${token}`,
        },
      });
      expect(response.status).toBe(200);
      expect(Array.isArray(response.data.runs)).toBe(true);
      expect(response.data.runs.length).toBe(0);
    });

    it('should return 401 or 403 if token is missing', async () => {
      const url = `${baseURL}${path}/${validProjectRef}/runs`;
      try {
        await axios.get(url); // no authorization header
        fail('Expected request to fail due to missing token');
      } catch (error) {
        const axiosError = error as AxiosError;
        expect([401, 403]).toContain(axiosError.response?.status);
      }
    });

    it('should return 401 or 403 if token is invalid', async () => {
      const url = `${baseURL}${path}/${validProjectRef}/runs`;
      try {
        await axios.get(url, {
          headers: {
            Authorization: 'Bearer invalid-token',
          },
        });
        fail('Expected request to fail due to invalid token');
      } catch (error) {
        const axiosError = error as AxiosError;
        expect([401, 403]).toContain(axiosError.response?.status);
      }
    });
  });
});